File: internal/router/embeddings.go
package router

import (
	"errors"
	"net/http"
	"time"

	"github.com/amaozhao/lazirouter/internal/combo"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/httpx"
	"github.com/amaozhao/lazirouter/internal/picker"
	"github.com/amaozhao/lazirouter/internal/quota"
	"github.com/amaozhao/lazirouter/internal/tenantctx"
	"github.com/amaozhao/lazirouter/internal/usage"
	"github.com/google/uuid"
)

func (d *Deps) handleEmbeddings(w http.ResponseWriter, r *http.Request) {
	started := time.Now()
	reqID := uuid.NewString()

	apiKey := httpx.BearerOrXAPIKey(r)
	if apiKey == "" {
		httpx.WriteError(w, errs.Auth("Missing Authorization header"))
		return
	}
	ctx, err := tenantctx.ResolveAPIKey(r.Context(), apiKey)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := tenantctx.EnforceTenantActive(ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := quota.Enforce(r.Context(), ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}

	body, err := httpx.ReadJSONMap(r)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	model, _ := body["model"].(string)
	if model == "" {
		httpx.WriteError(w, errs.Validation("Missing field: model"))
		return
	}
	if _, ok := body["input"]; !ok {
		httpx.WriteError(w, errs.Validation("Missing field: input"))
		return
	}

	attempts, err := combo.Resolve(r.Context(), ctx.TenantID, model)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}

	var errsList []map[string]any
	for _, attempt := range attempts {
		account, err := picker.Pick(r.Context(), d.Cfg.MasterKey, ctx.TenantID, attempt.Provider)
		if err != nil {
			var ae *errs.AppError
			if errors.As(err, &ae) && ae.Code == "no_account_available" {
				errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": ae.Message})
				continue
			}
			httpx.WriteError(w, err)
			return
		}
		baseURL := stringFrom(account.Metadata, "base_url")
		if baseURL == "" {
			baseURL = httpx.DefaultBaseURL(attempt.Provider)
		}
		upstreamKey := stringFrom(account.Credentials, "api_key")
		if baseURL == "" || upstreamKey == "" {
			errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": "connection missing base_url or api_key"})
			continue
		}

		upstreamBody := cloneMap(body)
		upstreamBody["model"] = attempt.UpstreamModel

		result, err := d.OpenAI.Embed(r.Context(), baseURL, upstreamKey, upstreamBody)
		ue := &usage.Event{
			TenantID: ctx.TenantID, APIKeyID: ctx.APIKeyID, ConnectionID: account.ConnectionID,
			Provider: attempt.Provider, Model: model, UpstreamModel: attempt.UpstreamModel,
			RequestID: reqID, Meta: map[string]any{"transport": "/v1/embeddings"},
		}
		if err != nil {
			var ae *errs.AppError
			status := 0
			if errors.As(err, &ae) {
				if sc, ok := ae.Meta["status_code"].(int); ok {
					status = sc
				}
			}
			retriable := status == 429 || (status >= 500 && status < 600) || !errors.As(err, &ae)
			if retriable && status > 0 {
				_, _ = picker.MarkCooldown(r.Context(), ctx.TenantID, attempt.Provider, account.ConnectionID)
			}
			_ = usage.Record(r.Context(), errEvent(ue, started, err))
			errsList = append(errsList, map[string]any{
				"step": attempt.Step, "provider": attempt.Provider, "status": status, "message": err.Error(),
			})
			if !retriable {
				httpx.WriteError(w, err)
				return
			}
			continue
		}
		prompt := toI64(asMap(result["usage"])["prompt_tokens"])
		pricing := usage.GetPricing(r.Context(), ctx.TenantID, attempt.Provider, attempt.UpstreamModel)
		ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
		ue.PromptTokens = prompt
		ue.CostMicros = usage.ComputeCost(pricing, prompt, 0)
		_ = usage.Record(r.Context(), ue)

		httpx.WriteJSON(w, http.StatusOK, result)
		return
	}

	httpx.WriteError(w, errs.NoAccountAvailable(map[string]any{"attempts": errsList}))
}

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return map[string]any{}
}

File: router/src/routes/embeddings.js
// POST /v1/embeddings — OpenAI-compatible embeddings entry point.
// Same pipeline as /v1/chat/completions but non-streaming only.
// Pricing for embedding models gets looked up via the same `pricing` table;
// callers that haven't seeded a row get cost=0.

import crypto from 'node:crypto';
import {
  ValidationError, UpstreamError, AuthError,
  NoAccountAvailableError, AppError,
} from '@lazirouter-cloud/shared';
import { enforceDailyQuota } from '@lazirouter-cloud/shared';
import { resolveApiKey, enforceTenantActive } from '../middleware/edgeAuth.js';
import { enforceRateLimit } from '../middleware/rateLimit.js';
import { pickAccount, markCooldown } from '../services/accountPicker.js';
import { recordUsage, getPricing, computeCost } from '../services/usage.js';
import { resolveAttempts } from '../services/combo.js';
import { embed } from '../providers/openaiCompatible.js';
import { readJson, defaultBaseUrl } from '../lib/http.js';

export async function handleEmbeddings(req, res) {
  const startedAt = Date.now();
  const requestId = crypto.randomUUID();

  const authHeader = req.headers['authorization'] || '';
  const apiKey = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : '';
  if (!apiKey) throw new AuthError('Missing Authorization header');
  const ctx = await resolveApiKey(apiKey);
  enforceTenantActive(ctx);
  await enforceRateLimit(ctx);
  await enforceDailyQuota(ctx);

  const body = await readJson(req);
  if (!body.model) throw new ValidationError('Missing field: model');
  if (body.input == null) throw new ValidationError('Missing field: input');

  const attempts = await resolveAttempts(ctx.tenantId, body.model);
  const errors = [];

  for (const attempt of attempts) {
    let account;
    try {
      account = await pickAccount(ctx.tenantId, attempt.provider);
    } catch (e) {
      if (e instanceof NoAccountAvailableError) {
        errors.push({ step: attempt.step, provider: attempt.provider, reason: e.message });
        continue;
      }
      throw e;
    }

    const baseUrl = account.metadata.base_url || defaultBaseUrl(attempt.provider);
    const upstreamApiKey = account.credentials.api_key;
    if (!baseUrl || !upstreamApiKey) {
      errors.push({ step: attempt.step, provider: attempt.provider, reason: 'connection missing base_url or api_key' });
      continue;
    }

    const upstreamBody = { ...body, model: attempt.upstreamModel };
    const usageBase = {
      tenantId: ctx.tenantId,
      apiKeyId: ctx.apiKeyId,
      connectionId: account.connectionId,
      provider: attempt.provider,
      model: body.model,
      upstreamModel: attempt.upstreamModel,
      requestId,
    };

    try {
      const result = await embed({ baseUrl, apiKey: upstreamApiKey, body: upstreamBody });
      const promptTokens = result.usage?.prompt_tokens || 0;
      const pricing = await getPricing(usageBase.tenantId, usageBase.provider, usageBase.upstreamModel);
      const costMicros = computeCost(pricing, promptTokens, 0);

      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify(result));

      await recordUsage({
        ...usageBase, status: 'ok', latencyMs: Date.now() - startedAt,
        promptTokens, completionTokens: 0, costMicros,
        meta: { transport: '/v1/embeddings' },
      });
      return;
    } catch (err) {
      const status = err instanceof UpstreamError ? err.meta?.status_code : null;
      const retriable = status === 429 || (status >= 500 && status < 600) || !(err instanceof AppError);
      if (retriable && status) await markCooldown(ctx.tenantId, attempt.provider, account.connectionId);

      await recordUsage({
        ...usageBase, status: 'error',
        latencyMs: Date.now() - startedAt,
        errorCode: status ? `upstream_${status}` : (err.code || 'unknown'),
        meta: { error: err.message, transport: '/v1/embeddings' },
      });
      errors.push({ step: attempt.step, provider: attempt.provider, status, message: err.message });
      if (!retriable) throw err;
    }
  }

  const allNoAccount = errors.length > 0 && errors.every(e => !e.status);
  if (allNoAccount) throw new NoAccountAvailableError({ attempts: errors });
  throw new UpstreamError('All upstream attempts failed', { attempts: errors });
}
